### PR TITLE
WIP: Improve performance for depwarn

### PR DESCRIPTION
**Pull request by omus**
During my investigation into why Appveyor CI was [consistently timing out](https://ci.appveyor.com/project/quinnj/timezones-jl/build/1.0.106) for TimeZones.jl on the nightly release of Julia I discovered that issue came from deprecation warnings.

Even though Julia avoids printing repeated deprecation warnings it will always compute the backtraces when encountering a deprecation. Running within a Windows VM `@time Pkg.build("TimeZones")` took 2073 seconds while this PR only takes 254 seconds (~8x improvement).

~~I personally don't think that a full stack trace is necessary for a deprecation warning so I think this is a reasonable solution. If others disagree~~ we could probably improve performance by only performing a backtrace once we know the deprecation warning will be displayed.

Remaining work for this PR includes:
- [ ] ~~Removing the `funcsym` parameter from `depwarn`~~
- [x] Test cases

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -58,40 +58,13 @@ end
...
-            end
-            if lkup.func == funcsym
-                @goto found
-            end
-        end
-    end
-    @label found
-    if i <= length(bt)
-        return bt[i]
+        warn(msg, once=True, key="$fn:$ln", filename=fn, lineno=ln)
```
Comment (yuyichao) [inline base/deprecated.jl]: true

**Comment by yuyichao** [comment]
I think the backtrace is very useful and wont tgis make depwarn print everytime?

**Comment by omus** [comment]
The deprecation warning is only printed once per file per line.

**Comment by omus** [comment]
Here is the difference in how things are displayed:

This PR:
```julia
WARNING: bytestring(p::Union{Ptr{Int8},Ptr{UInt8}},len::Integer) is deprecated, use String(p,len) instead.
while loading C:\msys64\home\omus\.julia\v0.5\WinRPM\src\WinRPM.jl, in expression starting on line 501
```

Previous behaviour:
```julia
WARNING: bytestring(p::Union{Ptr{Int8},Ptr{UInt8}},len::Integer) is deprecated, use String(p,len) instead.
 in depwarn(::String, ::Symbol) at .\deprecated.jl:64
 in bytestring(::Ptr{UInt8}, ::Int64) at .\deprecated.jl:50
 in cdata(::Ptr{Void}, ::Ptr{UInt8}, ::Int32) at C:\msys64\home\omus\.julia\v0.5\LibExpat\src\LibExpat.jl:166
 in XML_Parse(::Ptr{Void}, ::String, ::Int64, ::Int64) at C:\msys64\home\omus\.julia\v0.5\LibExpat\src\lX_common_h.jl:4
 in xp_parse(::String) at C:\msys64\home\omus\.julia\v0.5\LibExpat\src\LibExpat.jl:288
 in update(::Bool, ::Bool) at C:\msys64\home\omus\.julia\v0.5\WinRPM\src\WinRPM.jl:153
 in __init__() at C:\msys64\home\omus\.julia\v0.5\WinRPM\src\WinRPM.jl:41
 in include_from_node1(::String) at .\loading.jl:426
 in eval(::Module, ::Any) at .\boot.jl:225
 in require(::Symbol) at .\loading.jl:357
 in include_from_node1(::String) at .\loading.jl:426
 in eval(::Module, ::Any) at .\boot.jl:225
 in evalfile(::String, ::Array{String,1}) at .\loading.jl:442 (repeats 2 times)
 in cd(::##2#4, ::String) at .\file.jl:48
 in (::##1#3)(::IOStream) at .\none:13
 in open(::##1#3, ::String, ::String) at .\iostream.jl:113
 in eval(::Module, ::Any) at .\boot.jl:225
 in process_options(::Base.JLOptions) at .\client.jl:243
 in _start() at .\client.jl:322
while loading C:\msys64\home\omus\.julia\v0.5\WinRPM\src\WinRPM.jl, in expression starting on line 501
```

**Comment by mauro3** [comment]
Looks reasonable to me.

**Comment by yuyichao** [comment]
But you don't want to print only one depwarn per file per line since it is the toplevel file and line number, not where the code actually is....

**Comment by omus** [comment]
Hmm, I see what you mean now. I'll take another crack at this.

**Comment by JeffBezanson** [comment]
-1

I'm consistently surprised by how much work people are willing to put in to make living with deprecations more bearable, instead of just fixing the deprecations.

**Comment by omus** [comment]
@JeffBezanson I agree that deprecations should just be fixed but sometimes you need to live with them for a period of time until Compat is updated or a package you depend on is updated.

In this particular case `WinRPM` is generating a ton of calls to `depwarn` which is calling `backtrace` each time. The slowdown is so significant that Appveyor cancels the run after an hour which happens before it even starts building my package let alone running the tests.

Mostly this work came out of me trying to confirm that the deprecation warnings were responsible for the massive slow down.

**Comment by yuyichao** [comment]
I think it might be useful to expose a unwinder API to unwind a maximum number of frames. That should benefit the depwarn case as a side effect.

**Comment by tkelman** [comment]
backtraces are painfully slow on windows and the depwarn case is maybe a bit excessive, but you can see here that it looks like WinRPM wasn't even responsible for the bytestring call in question, LibExpat was. The line number in the short version wasn't even accurate.

**Comment by omus** [comment]
I've reworked things. I believe the behaviour is identical to how `depwarn` worked prior to this PR except that there is a pre-screening step which eliminates the need for computing a backtrace for each `depwarn` call. Looks like we can get the speed up with full backtraces, win-win.

@tkelman you are correct. LibExpat was the culprit not WinRPM.

File: base/deprecated.jl [hunk 2]
```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
...
+const have_dep_warned = Set()
+
 function depwarn(msg, funcsym)
     opts = JLOptions()
-    if opts.depwarn > 0
-        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+    if opts.depwarn == 1  # raise a warning
         fn = String(unsafe_load(cglobal(:jl_filename, Ptr{Cchar})))
+        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+        key = hash(string(fn, ln, msg))
```
Comment (tkelman) [inline base/deprecated.jl]: isn't this double-hashing? why not just put the string in the set?

File: base/deprecated.jl [hunk 3]
```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
...
+        (key in have_dep_warned) && return
```
Comment (yuyichao) [inline base/deprecated.jl]: Doesn't this still turn off all the warnings for the same depwarn at different locations in the code called from the same toplevel line?

File: base/deprecated.jl (hunk 2 above)
Comment (omus) [inline base/deprecated.jl]: I could just put the string in the set but I thought hashing would be a more memory efficient approach.

I haven't messed with Julia's hashing that much. What's the best way to hash the subcomponents together?

File: base/deprecated.jl (hunk 3 above)
Comment (omus) [inline base/deprecated.jl]: @yuyichao I was wondering about than myself. In practice the `have_dep_warned` filter seems less granular than the caller filter below.

My ad hoc testing of comparing the before and after output has shown that the same amount of deprecation warnings are emitted. 

So far everything is just proof-of-concept and I'll need to figure out some good tests. Could you provide me with a test that would cover your concern?

Comment (yuyichao) [inline base/deprecated.jl]:
```jl
@noinline function f1()
    Base.depwarn("1", :f)
end
@noinline function f2()
    Base.depwarn("1", :f)
end
@noinline function f3()
    f1()
    f2()
end
f3()
```

Comment (yuyichao) [inline base/deprecated.jl]: In general, using the filename and line number as keys is just wrong. They can be different for the same usage in the code and they can be the same for multiple usage calling from the same toplevel line.
The only solution I can think of is https://github.com/JuliaLang/julia/pull/16617#issuecomment-222231575

File: base/deprecated.jl [hunk 4]
```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
...
-        i += 1
         for lkup in lkups
-            if lkup === StackTraces.UNKNOWN
-                continue
-            end
-            if lkup.func == funcsym
-                @goto found
-            end
+            lkup === StackTraces.UNKNOWN && continue
+            lkup.func == funcsym && return bt[i]
```
Comment (carnaval) [inline base/deprecated.jl]: Isn't this broken ? It used to return the frame containing the call to the deprecated function and it looks like it's returning the deprecated frame itself.

Comment (omus) [inline base/deprecated.jl]: I messed up here. Thanks for catching the mistake.

**Comment by omus** [comment]
I've added a new backtrace method `backtrace(limit::Integer)` which allows users to limit the amount of Julia functions that are unwinded. This method isn't as fast as what I had before but works correctly for  the issue @yuyichao pointed out. On the original problem I mentioned in the description it takes ~700 seconds to run making it ~3x faster.

Additionally, I have updated `julia_cmd` to pass through the `--depwarn` option which allows developers to disable deprecation warnings from within CI configuration files.

File: base/deprecated.jl [hunk 5]
```diff
@@ -58,24 +58,22 @@ end
...
+        throw(ErrorException(msg))
     end
 end
 
 function firstcaller(bt::Array{Ptr{Void},1}, funcsym::Symbol)
     # Identify the calling line
     i = 1
-    while i <= length(bt)
+    len = length(bt)
+    while i <= len
```
Comment (tkelman) [inline base/deprecated.jl]: the compiler can't figure this one out?

**Comment by omus** [comment]
The method of limiting the amount of unwinding in a backtrace does seem to improve performance. Unfortunately I have noticed that this performance improvement is inconsistent and sometimes the PR is just as slow as the original code.

I would definitely like if someone could look over the `src/stackwalk.c` changes as I may have missed something in that section of code.

File: src/stackwalk.c [hunk 6]
```diff
@@ -84,8 +84,28 @@ size_t rec_backtrace(uintptr_t *data, size_t maxsize)
     return rec_backtrace_ctx(data, maxsize, &context);
 }
 
+int jl_num_frames(uintptr_t ip, int skipC)
```
Comment (yuyichao) [inline src/stackwalk.c]: This seems expensive. I don't think this is needed.

Comment (omus) [inline src/stackwalk.c]: Prior to adding this it was difficult to estimate how much of the backtrace was from C or from Julia. I'm open to ideas though.

File: base/deprecated.jl (hunk 5 above)
Comment (vtjnash) [inline base/deprecated.jl]: lengths of arrays aren't constant, although loading this value from the stack spill vs. loading it from bt shouldn't be different in performance

File: src/stackwalk.c (hunk 6 above)
Comment (vtjnash) [inline src/stackwalk.c]: I'd be open to having a copy of `firstcaller` written in C. I'm not really interested in simply truncating backtraces.

Comment (yuyichao) [inline src/stackwalk.c]: I mainly mean that this is an internal API and you don't need to estimate it at all.

Comment (omus) [inline src/stackwalk.c]: Should I just do `firstcaller` or maybe something like caller plus n-ancestors?

File: base/deprecated.jl (hunk 5 above)
Comment (tkelman) [inline base/deprecated.jl]: it's constant if nothing changes it, though proving that is likely hard

**Comment by omus** [comment]
I took @vtjnash's suggestion and made a version of `firstcaller` written in C: `jl_backtrace_caller` (working name).

File: src/stackwalk.c [hunk 7]
```diff
@@ -84,6 +84,84 @@ size_t rec_backtrace(uintptr_t *data, size_t maxsize)
...
+    free(frames);
+    return found;
+}
+
+JL_DLLEXPORT jl_value_t *jl_backtrace_caller(jl_sym_t *name, int num_callers)
+{
+    if (num_callers < 0)
+        return NULL;
+
+    jl_value_t *ip = (jl_value_t*)jl_svec2(jl_voidpointer_type, jl_box_long(1));
```
Comment (omus) [inline src/stackwalk.c]: I'm rather unsure if I am allocating this correctly

Comment (vtjnash) [inline src/stackwalk.c]: are you trying to allocate a `Ptr{Void}` box? that would be impossible, however. this should simply be a C variable, no need to involve the julia gc here (`ip` isn't a `jl_value_t*` either):
```
uintptr_t ip;
```

Comment (omus) [inline src/stackwalk.c]: Thanks, seems obvious now. I think I was worried that if I didn't allocate memory with julia that I could get left with a dangling pointer.

**Comment by musm** [comment]
@omus hate to bump up such an old PR, but  the contents of this PR are so stale I think it might be best to close this PR as it might be harder to rebase than  open up a new PR with equivalent changes.

Thus, I'm going to close this, but if you feel like that was a mistake, please go ahead and re-open.

**Comment by omus** [comment]
> Thus, I'm going to close this, but if you feel like that was a mistake, please go ahead and re-open.

Totally fair. I'd like to see deprecation warnings be less expensive but I agree I'd just start a new PR at this point.

**Pull request closed without merging**